sys-kernel/ugrd/ugrd-1.19.1.ebuild

```diff
@@ -46,3 +53,23 @@ pkg_postinst() {
 	optfeature "ugrd.crypto.gpg support" app-crypt/gnupg
 	optfeature "ugrd.fs.lvm support" sys-fs/lvm2[lvm]
 }
+
+distutils_enable_tests unittest
+
+src_test() {
+	addwrite /dev/kvm
+	if [[ ! -w '/dev/kvm' ]]; then
+		ewarn "Cannot write to /dev/kvm, skipping tests"
+		return 1
+	fi
+	if [[ ! -r "$(command -v mount)" ]]; then
+		ewarn "Cannot read the mount binary, tests may fail until"
+		ewarn "util-linux is re-emerged without the sfperms feature."
+	fi
```

Comment:
I'm not sure about this. In `kernel-install.eclass` we also use qemu to test the to-be-installed kernel plus a dracut initramfs, but AFAIK we don't need to `addwrite /dev/kvm` to do so.  Why do we need to do this here, and isn't it possible to do something more like `kernel-install.eclass`?

Reply:
> I'm not sure about this. In `kernel-install.eclass` we also use qemu to test the to-be-installed kernel plus a dracut initramfs, but AFAIK we don't need to `addwrite /dev/kvm` to do so. Why do we need to do this here, and isn't it possible to do something more like `kernel-install.eclass`?

like which part of it? it's just meant to be a simple test that the portage user can run kvm stuff, i was getting permission errors and I think checking that it can write there is a good test.

Reply:
I'm just confused why this explicitly needs write access to `/dev/kvm` when `kernel-install.eclass` does not.

Reply:
Hmm, the dracut ebuild also needs this to execute the tests apparently:
```
src_test() {
	if [[ ${EUID} != 0 ]]; then
		# Tests need root privileges, bug #298014
		ewarn "Skipping tests: Not running as root."
	elif [[ ! -w /dev/kvm ]]; then
		ewarn "Skipping tests: Unable to access /dev/kvm."
	else
		emake -C test check
	fi
}
```

Maybe we can structure this the same way here:

```
src_test() {
	if [[ ${EUID} != 0 ]]; then
		# Tests need root privileges, bug #298014
		ewarn "Skipping tests: Not running as root."
	elif [[ ! -w /dev/kvm ]]; then
		ewarn "Skipping tests: Unable to access /dev/kvm."
	else
		distutils-r1_src_test
	fi
}
```

 Though I still don't understand why `kernel-install.eclass` does not need this.

Reply:
Are you sure it doesn't? I have not tried running those tests, have just read the code, but I couldn't get portage to properly run the tests until i added portage to the kvm/qemu groups. If there is some other way to properly execute qemu-based tests as the portage user, I'd like to use that instead of checking and expecting certain group config.

Reply:
it may be that the kernel-install tests are using qemu without `enable-kvm`?

Reply:
> Are you sure it doesn't?

Not really, I'm a bit confused now. On my system gentoo-kernel tests work and I do not remember doing anything special to make them work, nor is my portage user part of the kvm or qemu groups.

Reply:
my tests fail without using kvm: `'[    1.882941] traps: init[1] trap invalid opcode ip:7f9663305d99 sp:7ffded3bf410 error:0 in ld-linux-x86-64.so.2[7f96632e6000+29000]'` maybe because my stuff is build with `march=native`?

Reply:
dracut tests fail as well:
```
[    2.395864] traps: init[1] trap invalid opcode ip:7f61636ded99 sp:7fff5f073b10 error:0 in ld-linux-x86-64.so.2[7f61636bf000+29000]
[    2.396606] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000004
```

So I think kvm is required for tests to pass (at least on my system)

Reply:
also fwiw ugrd tests do not need root, just kvm access, and I think only to do a proper "host" test, which is ideal for ugrd since it's designed to make images for a particular host. I think using qemu without kvm will break most testing.

Reply:
> also fwiw ugrd tests do not need root, just kvm access, 

Does the check for root not effectively do the same as your check for sfperms? It's simpler to re-run the test phase with root privileges then it is to re-compile another package with a specific feature disabled.

In any case, I'd prefer to not `addwrite` and instead simply warn and skip the tests if we can't access `/dev/kvm` (like in the dracut ebuild).

Reply:
See `PROPERTIES="test_privileged"` and `ALLOW_TEST=privileged` in `man make.conf` (we should probably use that in the dracut ebuild too)

Reply:
> This makes sense. Do you have thoughts on the kvm requirement? `cpu max` should tell qemu to enable all cpu features, but I think this may not work with some very modern cpus, as I couldn't get Dracut or ugrd to start without KVM.

I don't really mind the kvm requirement, it's inconvenient but the dracut ebuild has it too so there is precedent for this. I'm hesitant about the `addwrite` though, it silently opens up the sandbox a bit, and all the documentation we have on this says not to use it if it can be avoided. I think I'd prefer to warn and skip the tests if we can't write to `/dev/kvm`, and leave it up to the user to either execute the tests privileged or add the `portage` user to the `kvm` group. The new `PROPERTIES="test_privileged"` was introduced relatively recently precisely for cases such as this: https://bugs.gentoo.org/924585.

Reply:
> > This makes sense. Do you have thoughts on the kvm requirement? `cpu max` should tell qemu to enable all cpu features, but I think this may not work with some very modern cpus, as I couldn't get Dracut or ugrd to start without KVM.
> 
> I don't really mind the kvm requirement, it's inconvenient but the dracut ebuild has it too so there is precedent for this. I'm hesitant about the `addwrite` though, it silently opens up the sandbox a bit, and all the documentation we have on this says not to use it if it can be avoided. I think I'd prefer to warn and skip the tests if we can't write to `/dev/kvm`, and leave it up to the user to either execute the tests privileged or add the `portage` user to the `kvm` group. The new `PROPERTIES="test_privileged"` was introduced relatively recently precisely for cases such as this: https://bugs.gentoo.org/924585.

Ok, so it could have some `if -w /dev/kvm` type test, I could possibly make it pass `test_cpu max` in that case, so it doesn't try to use kvm, for cases where qemu tests could work. I think the main issue is running the tests using binaries built for the host, where QEMU won't properly emulate all required instructions.

`PROPERTIES="test_privileged"` will make the tests run as root, right? I don't think this should really be necessary, but it may make sense if sfperms is still the default. I would like to keep the tests running with the lest privilege required. It seems that adding portage to the kvm group may be the best way of doing that? It would be interesting if there were some mechanism to give portage temporary perms while it runs tests or something, but I can see how this could be challenging/breakable.

Reply:
> `PROPERTIES="test_privileged"` will make the tests run as root, right?

Yes

> It seems that adding portage to the kvm group may be the best way of doing that? It would be interesting if there were some mechanism to give portage temporary perms while it runs tests or something, but I can see how this could be challenging/breakable.

Well that is basically what `PROPERTIES="test_privileged"` does, albeit in a very all-or-nothing brute force kind of way. It would indeed be nice to have a somewhat finer control, though I suppose `addwrite` is one way of having finer control.

I'm honestly not sure what is best here, maybe I'll ask around for some more opinions.

Reply:
> > `PROPERTIES="test_privileged"` will make the tests run as root, right?
> 
> Yes
> 
> > It seems that adding portage to the kvm group may be the best way of doing that? It would be interesting if there were some mechanism to give portage temporary perms while it runs tests or something, but I can see how this could be challenging/breakable.
> 
> Well that is basically what `PROPERTIES="test_privileged"` does, albeit in a very all-or-nothing brute force kind of way. It would indeed be nice to have a somewhat finer control, though I suppose `addwrite` is one way of having finer control.
> 
> I'm honestly not sure what is best here, maybe I'll ask around for some more opinions.

Thanks for the thoughts/feedback so far. Maybe this could be used as an opportunity to tighten up the dracut tests too. It would be interesting if dist-kernel had a more generic kernel test mechanism, where it simply made its own vm and attempted any/all initrd generators. I'm fine maintaining a test suite within ugrd, but it would be interesting if portage were able to test all available initramfs options at once.

Reply:
So I asked around a bit, and there is no established policy about `addwrite` versus `PROPERTIES=test_privileged`, but the more I think about it the more I think there really should be.

How about this, for now we do it the way it's done in the dracut ebuild (so no `addwrite`, nor `PROPERTIES=test_privileged`, just check and skip if conditions are not met), then we can at least get this version bump in. Then after, we can discuss on the mailing list about what the proper solution should be, maybe implement that as policy, and then fix this both here and in the dracut ebuild.